### PR TITLE
refactor(p2p)!: Extract TCP transport

### PR DESCRIPTION
**Pull request by melekes**
Closes https://github.com/cometbft/cometbft/issues/4301

This PR refactors the peer-to-peer (p2p) module by extracting the TCP transport into its own component, laying groundwork for potential QUIC transport integration. It introduces a breaking change to the Go API, particularly in the handling of peer connections and transport interfaces.

**Important: no logic has been changed; just moving stuff and updating interfaces.**

# Changes

## Public API

* Updated `Transport` interface

```go
// Transport emits and connects to Peers. The implementation of Peer is left to
// the transport. Each transport is also responsible to filter establishing
// peers specific to its domain.
type Transport interface {
	// NetAddr returns the network address of the local node.
	NetAddr() na.Addr

	// Accept waits for and returns the next connection to the local node.
	Accept() (net.Conn, *na.Addr, error)

	// Dial dials the given address and returns a connection.
	Dial(addr na.Addr) (net.Conn, error)

	// Cleanup any resources associated with the given connection.
	//
	// Must be run when the peer is dropped for any reason.
	Cleanup(conn net.Conn) error
}
```

## New Go packages (extracted from `p2p`)

* `netaddress` - network address
* `nodeinfo` - node's info
*  `nodekey` - ID of the node
*  `transport/tcp` - TCP transport

## Internalized Go packages

* `fuzz` - fuzzing connection

#go-api-breaking

### CODE REVIEW

**Comment by cason** [comment]
A general fast comment here: if we don't rename `GetChannels()` reactor's method and the type that stores it, we have much less changes and maybe is not (that) API breaking?

**Comment by melekes** [comment]
> A general fast comment here: if we don't rename `GetChannels()` reactor's method and the type that stores it, we have much less changes and maybe is not (that) API breaking?

It will be API-breaking anyway because I had to extract many packages. But if you think the renaming is unjustified, I can revert it to the "channels" concept.

File: p2p/transport/tcp/tcp.go [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+	// TODO(xla): Evaluate if we should apply filters if we explicitly dial.
+	if err := mt.filterConn(c); err != nil {
+		return nil, err
+	}
+
+	secretConn, err := mt.upgrade(c, &addr)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (andynog) [inline p2p/transport/tcp/tcp.go:193]: just noticed that in the previous file, there was a cfg parameter and the `outbound` was set. I'm assuming this is not needed or was removed on purpose but just wanted to highlight here
```
cfg.outbound = true
```

**Review by andynog — APPROVED** [review]
Overall looks good to me. Made some comments and suggestions. Also ran a few tests. Good job. 👍

File: internal/consensus/reactor.go [hunk 2]
```diff
@@ -151,41 +152,41 @@ conR:
 	}
 }
 
-// GetChannels implements Reactor.
-func (*Reactor) GetChannels() []*p2p.ChannelDescriptor {
+// StreamDescriptors implements Reactor.
```
[the remark this answers is not in the document]
Comment (melekes) [reply internal/consensus/reactor.go:155]: So, "channel" is what we were using before, but I don't know where this naming concept is coming from. Channels are not mentioned anywhere in TCP or multiplexed TCP.

Where's in QUIC, there's a direct concept of data streams. Hence, I've decided to change the name.

File: internal/blocksync/reactor.go [hunk 3]
```diff
@@ -170,16 +172,16 @@ func (bcR *Reactor) OnStop() {
 	}
 }
 
-// GetChannels implements Reactor.
-func (*Reactor) GetChannels() []*p2p.ChannelDescriptor {
-	return []*p2p.ChannelDescriptor{
-		{
+// StreamDescriptors implements Reactor.
+func (*Reactor) StreamDescriptors() []p2p.StreamDescriptor {
```
[the remark this answers is not in the document]
Comment (melekes) [reply internal/blocksync/reactor.go:176]: It does

File: p2p/peer.go [hunk 4]
```diff
@@ -444,11 +445,56 @@ func createMConnection(
 		onPeerError(p, r)
 	}
 
-	return cmtconn.NewMConnectionWithConfig(
+	// filter out non-tcpconn.ChannelDescriptor streams
+	chDescs := make([]*tcpconn.ChannelDescriptor, 0, len(streams))
```
[the remark this answers is not in the document]
Comment (melekes) [reply p2p/peer.go:449]: This part will be refactored out later :)

File: spec/p2p/reactor-api/reactor.qnt [hunk 5]
```diff
@@ -77,10 +77,10 @@ module reactor {
 	// Pure definitions below represent the `p2p.Reactor` interface methods:
 
 	/*
-	 * GetChannels returns the list of MConnection.ChannelDescriptor. Make sure
+	 * StreamDescriptors returns the list of MConnection.ChannelDescriptor. Make sure
```
[the remark this answers is not in the document]
Comment (melekes) [reply spec/p2p/reactor-api/reactor.qnt]: I would rather not touch quint spec.

**Comment by tac0turtle** [comment]
nice job!!

**Pull request merged**